### ISSUE TITLE
"Build Project" does nothing right after importing a project in a fresh workspace

### ISSUE THREAD

**Issue by AdrianOltean** [eclipse-cdt/cdt#323]
**Describe the bug**
"Build Project" does nothing right after importing a project in a fresh workspace

**To Reproduce**
1. Install Eclipse C/C++ 2020-12 or newer (e.g. 2022-12) Maybe problem can be reproduced with older build as well but I don't have others installed. Moreover, I only tried on Windows 10 hosts.
2. Create a basic Hello World C example. 
3. Delete (only from workspace, not from disk) previously created project.
4. Import again the project using the shortcut from Project Explorer.
5. Once imported, select project in Project Explorer.
6. Go to Project => Build Project.
Note: I usually have a project that I want to import, so I don't do the steps from 2 & 3, instead I simply import the project in a clean workspace.

**Expected behavior**
Build process must be initiated. Actual result: nothing happens.

**Desktop (please complete the following information):**
 - OS: Windows 10 hosts
 - Version: Eclipse C/C++ IDE 2020-12 and newer

**Additional context**
I don't fully understand the code around this but it looks like selection from Project Explorer is empty while checking enablement of `CDTBuildAction`. See screenshot.
![image](https://user-images.githubusercontent.com/127829751/224932046-b4d2d2ab-f633-43f2-9a6d-2070ed9cc019.png)

Some relevant stack frames:
```
Thread [main] (Suspended)	
	BuildUtilities.findSelectedProjects(IWorkbenchWindow) line: 102	
	BuildGroup$CDTBuildAction(BuildAction).isEnabled() line: 231	
	RetargetActionWithDefault(RetargetAction).setActionHandler(IAction) line: 268	
	RetargetActionWithDefault.setActionHandler(IAction) line: 46	
	RetargetActionWithDefault(RetargetAction).propagateChange(PropertyChangeEvent) line: 204	
	0x0000000800694000.propertyChange(PropertyChangeEvent) line: not available	
	ActionBars(SubActionBars).firePropertyChange(PropertyChangeEvent) line: 289	
	ActionBars(SubActionBars).fireActionHandlersChanged() line: 273	
	ActionBars(SubActionBars).updateActionBars() line: 603	
	ActionBars.updateActionBars() line: 88	
	NavigatorActionService.fillActionBars(IActionBars) line: 274	
	CommonNavigatorManager.selectionChanged(SelectionChangedEvent) line: 220	
	Viewer$1.run() line: 151	
	SafeRunner.run(ISafeRunnable) line: 45	
	0x000000080025a408.run(ISafeRunnable) line: not available	
	SafeRunnable.run(ISafeRunnable) line: 174	
	CommonViewer(Viewer).fireSelectionChanged(SelectionChangedEvent) line: 148	
	CommonViewer(StructuredViewer).updateSelection(ISelection) line: 2128	
	CommonViewer(StructuredViewer).handleSelect(SelectionEvent) line: 1176	
	CommonViewer.handleSelect(SelectionEvent) line: 466	
	StructuredViewer$4.widgetSelected(SelectionEvent) line: 1205	
	OpenStrategy.fireSelectionEvent(SelectionEvent) line: 263	
...
```

Then, the same `isEnabled()` call ends-up executing `getSelectedResources()` which finds no item selected (even though there's a project selected in Project Explorer). It also sets `selectionDirty=false`, meaning that other calls to `getSelectedResources()` will not updated internally cached selection, until something else gets selected in Project Explorer.
![image](https://user-images.githubusercontent.com/127829751/224933637-fbea3206-1e86-485b-8d39-292b20944b35.png)
Some relevant stack frames:
```
Thread [main] (Suspended (breakpoint at line 195 in SelectionListenerAction))	
	BuildGroup$CDTBuildAction(SelectionListenerAction).getSelectedResources() line: 195	
	BuildGroup$CDTBuildAction(BuildAction).getBuildConfigurationsToBuild() line: 188	
	BuildGroup$CDTBuildAction(BuildAction).getProjectsToBuild() line: 169	
	BuildGroup$CDTBuildAction(BuildAction).updateSelection(IStructuredSelection) line: 318	
	BuildGroup$CDTBuildAction.updateSelection(IStructuredSelection) line: 69	
	BuildGroup$CDTBuildAction(BaseSelectionListenerAction).selectionChanged(IStructuredSelection) line: 128	
	BuildGroup$CDTBuildAction(BuildAction).isEnabled() line: 231	
	RetargetActionWithDefault(RetargetAction).setActionHandler(IAction) line: 268	
	RetargetActionWithDefault.setActionHandler(IAction) line: 46	
	RetargetActionWithDefault(RetargetAction).propagateChange(PropertyChangeEvent) line: 204	
	0x0000000800694000.propertyChange(PropertyChangeEvent) line: not available	
	ActionBars(SubActionBars).firePropertyChange(PropertyChangeEvent) line: 289	
	ActionBars(SubActionBars).fireActionHandlersChanged() line: 273	
	ActionBars(SubActionBars).updateActionBars() line: 603	
	ActionBars.updateActionBars() line: 88	
	NavigatorActionService.fillActionBars(IActionBars) line: 274	
	CommonNavigatorManager.selectionChanged(SelectionChangedEvent) line: 220	
	Viewer$1.run() line: 151	
	SafeRunner.run(ISafeRunnable) line: 45	
	0x000000080025a408.run(ISafeRunnable) line: not available	
	SafeRunnable.run(ISafeRunnable) line: 174	
	CommonViewer(Viewer).fireSelectionChanged(SelectionChangedEvent) line: 148	
	CommonViewer(StructuredViewer).updateSelection(ISelection) line: 2128	
	CommonViewer(StructuredViewer).handleSelect(SelectionEvent) line: 1176	
...
```

**Comment by AdrianOltean** [comment]
Maybe there's someone with a better understanding of the code I referenced above and finds an easy fix...

Thanks,
Adrian

**Comment by jantje** [comment]
I have seen similar behaviour which I could work around by doing a clean first.
In my understanding CDT (managed build?) caches the build state. Once it is in the "I'm build" but out of sync with reality (due to a unexpected use case or bug) trying to build will not fix it.
Changing code (inclusive saving); deleting the build folder or running a clean are ways I know of to reset the "I'm build" flag.
I hope this helps.

**Comment by AdrianOltean** [comment]
Thanks. Another workaround would be to change selection in Project Explorer and then move it back to the desired project...

**Comment by ewaterlander** [comment]
I have seen this too with several versions of CDT, on Linux. Especially if you create the first project and immediately after that unfold the the project in the Project Explorer. Eclipse freezes. The only way out is to kill it. Waiting  a while before unfolding helps.

**Comment by jonahgraham** [comment]
> Eclipse freezes. 

If you can ever reproduce this, can you save the stack traces (e.g. using jstack)?

**Comment by ewaterlander** [comment]
> > Eclipse freezes.
> 
> If you can ever reproduce this, can you save the stack traces (e.g. using jstack)?

Yes, I will.

**Comment by ewaterlander** [comment]
I can reproduce it with my own plugin that is built upon Core Build Makefile.  What I see with jstack is that several processes are getting blocked on getting a lock on the 'configs' variable in CBuildConfigurationManager.getBuildConfiguration(IBuildConfiguration buildConfig).
I think it's another problem than the original posted here, but it also only happens with the first project in a fresh workspace.

**Comment by ewaterlander** [comment]
For me the problem goes away if I disable Project->Build Automatically.
The indexer is involved in the problem. Not all source files can be compiled with the default gcc toolchain in my case.

**Comment by jonahgraham** [comment]
Build automatically is off for most ISVs using CDT - see the following for a relatively recent discussion on this topic:

- [Bug 571849](https://bugs.eclipse.org/bugs/show_bug.cgi?id=571849)
- "Look at build automatically in C/C++ EPP" entry in "CDT Monthly Call March 10 2021" [minutes](https://hackmd.io/@jonahgraham/cdt-call-mar-10-2021)
- The [setting change](https://github.com/eclipse-packaging/packages/blob/93b5963729d7144705f935e52b652ffd74c1f01c/packages/org.eclipse.epp.package.cpp/plugin_customization.ini#L65-L67) that products based on CDT often do, including what Eclipse IDE for C/C++ does

**Comment by ewaterlander** [comment]
> Build automatically is off for most ISVs using CDT - see the following for a relatively recent discussion on this topic:

Thanks!

**Comment by ewaterlander** [comment]
Unfortunately the problem is not gone with Build Automatically disabled, it only happens less frequent. I don't want to hijack this issue, so I created a new issue. See issue #424

**Comment by jantje** [comment]
I ran into this problem again today and I think it is really annoying to select "build" in the user interface and nothing happens.
I mean the end user requested a build and something decides it is smarter than the end user and refuses to build.
If that something is really smarter that is fine for me but if it "fails to be smarter in some cases" it is very annoying. It feels like "stupid has the power and stupid wants you to know it has the power". 😭 

Going through the clean process -as I described above- is only a feasible workaround in projects that do not take long to build. But for these project (that build really fast) it is nearly impossible to see whether CDT actually build something or ignored the end user request). I have been thinking CDT did the wrong thing but it actually didn't do anything and I was looking at old data. 🤡 

my 2 cent:
Personally I only see a valid reason to refuse to build in autobuild mode (or other automagically triggered builds).
End user requests (No matter how stupid they may seem to be) should be executed.
And if you do not agree on this: can we have a "clean build status" or "force build" in the gui?

**Comment by jantje** [comment]
> my 2 cent:
> Personally I only see a valid reason to refuse to build in autobuild mode (or other automagically triggered builds).
> End user requests (No matter how stupid they may seem to be) should be executed.
> And if you do not agree on this: can we have a "clean build status" or "force build" in the gui?

Seem I was wrong here, sorry for tat. I just learned that it is not really CDT who is holding the build command (it is a setting in the builder extensionpoint but turning it to always build does not result in a hallelujah experience in a C++ environment). So please forget the request.
